File: sdk/android/story-sdk/src/main/kotlin/com/openstory/sdk/internal/cache/ViewedStoryStateSnapshot.kt
package com.openstory.sdk.internal.cache

import com.openstory.sdk.internal.network.SdkFeedGroupPayload
import com.openstory.sdk.internal.network.SdkFeedStoryPayload

internal data class ViewedStoryStateSnapshot(
    val viewedStoryRevisionIds: Set<String>,
) {
    fun isStoryViewed(story: SdkFeedStoryPayload): Boolean {
        return viewedStoryRevisionIds.contains(story.revisionId)
    }

    fun isGroupViewed(group: SdkFeedGroupPayload): Boolean {
        return group.stories.isNotEmpty() && group.stories.all(::isStoryViewed)
    }

    fun firstUnviewedStoryIndex(group: SdkFeedGroupPayload): Int {
        val firstUnviewedIndex = group.stories.indexOfFirst { story -> !isStoryViewed(story) }
        return if (firstUnviewedIndex >= 0) {
            firstUnviewedIndex
        } else {
            0
        }
    }

    fun withViewedStory(storyRevisionId: String): ViewedStoryStateSnapshot {
        if (storyRevisionId.isBlank() || viewedStoryRevisionIds.contains(storyRevisionId)) {
            return this
        }

        return copy(viewedStoryRevisionIds = viewedStoryRevisionIds + storyRevisionId)
    }

    companion object {
        val EMPTY = ViewedStoryStateSnapshot(emptySet())
    }
}

File: sdk/flutter/ios/open_story_flutter/Sources/open_story_flutter/OpenStorySDK/Internal/Cache/ViewedStoryStateSnapshot.swift
import Foundation

internal struct ViewedStoryStateSnapshot: Equatable, Sendable {
    let viewedStoryRevisionIds: Set<String>

    func isStoryViewed(_ story: SdkFeedStoryPayload) -> Bool {
        viewedStoryRevisionIds.contains(story.revisionId)
    }

    func isGroupViewed(_ group: SdkFeedGroupPayload) -> Bool {
        !group.stories.isEmpty && group.stories.allSatisfy(isStoryViewed(_:))
    }

    func firstUnviewedStoryIndex(in group: SdkFeedGroupPayload) -> Int {
        if let firstUnviewed = group.stories.firstIndex(where: { !isStoryViewed($0) }) {
            return firstUnviewed
        }
        return 0
    }

    func withViewedStory(_ storyRevisionId: String) -> ViewedStoryStateSnapshot {
        let trimmed = storyRevisionId.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty, !viewedStoryRevisionIds.contains(trimmed) else {
            return self
        }

        return ViewedStoryStateSnapshot(
            viewedStoryRevisionIds: viewedStoryRevisionIds.union([trimmed])
        )
    }

    static let empty = ViewedStoryStateSnapshot(viewedStoryRevisionIds: [])
}
